rama-http-types/src/body/sse/event_stream.rs

```diff
@@ -0,0 +1,673 @@
+        buffer: String,
+        builder: EventBuilder<T>,
+        state: EventStreamState,
+        last_event_id: Option<SmolStr>,
+        _event_data: PhantomData<fn() -> T>,
+    }
+}
+
+impl<S, T> EventStream<S, T> {
+    /// Initialize the EventStream with a Stream
+    pub fn new(stream: S) -> Self {
+        Self {
+            stream: Utf8Stream::new(stream),
+            buffer: String::new(),
+            builder: EventBuilder::default(),
+            state: EventStreamState::NotStarted,
+            last_event_id: None,
+            _event_data: PhantomData,
+        }
+    }
+
+    /// Set the last event ID of the stream. Useful for initializing the stream with a previous
+    /// last event ID
+    pub fn try_set_last_event_id(&mut self, id: impl AsRef<str>) -> Result<(), OpaqueError> {
+        let id = id.as_ref();
+        if id.contains(['\n', '\r', '\0']) {
+            return Err(OpaqueError::from_std(EventBuildError::invalid_characters(
+                id,
+            )));
+        }
+        self.last_event_id = Some(SmolStr::new(id));
+        Ok(())
+    }
+
+    /// Get the last event ID of the stream
+    pub fn last_event_id(&self) -> Option<&str> {
+        self.last_event_id.as_deref()
+    }
+}
+
+fn parse_event<T: EventDataRead>(
+    buffer: &mut String,
+    builder: &mut EventBuilder<T>,
+) -> Result<Option<Event<T>>, OpaqueError> {
+    if buffer.is_empty() {
+        return Ok(None);
+    }
+    loop {
+        match line(buffer.as_ref()) {
+            Ok((rem, next_line)) => {
+                builder.add(next_line);
+                let consumed = buffer.len() - rem.len();
+                let rem = buffer.split_off(consumed);
+                *buffer = rem;
+                if builder.is_complete {
+                    return builder.try_dispatch().map(Some);
+                }
+            }
+            Err(nom::Err::Incomplete(_)) => {
+                return Ok(None);
+            }
+            Err(nom::Err::Error(err) | nom::Err::Failure(err)) => {
+                return Err(OpaqueError::from_display(format!("SSE parse error: {err}")));
+            }
+        }
+    }
+}
+
+impl<S, B, E, T> Stream for EventStream<S, T>
+where
+    S: Stream<Item = Result<B, E>>,
+    E: Into<BoxError>,
+    B: AsRef<[u8]>,
+    T: EventDataRead,
+{
+    type Item = Result<Event<T>, OpaqueError>;
+
+    fn poll_next(self: Pin<&mut Self>, cx: &mut Context) -> Poll<Option<Self::Item>> {
+        let mut this = self.project();
+
+        match parse_event(this.buffer, this.builder) {
+            Ok(Some(event)) => {
+                *this.last_event_id = event.id().map(SmolStr::new);
+                return Poll::Ready(Some(Ok(event)));
+            }
+            Err(err) => return Poll::Ready(Some(Err(err))),
+            _ => {}
+        }
+
+        if this.state.is_terminated() {
+            return Poll::Ready(None);
+        }
+
+        loop {
+            match this.stream.as_mut().poll_next(cx) {
```

Comment:
```suggestion
            match ready!(this.stream.as_mut().poll_next(cx)) {
```
ready! macro makes this quite a lot nicer to work with